ReactAndroid/src/main/java/com/facebook/react/ReactActivity.java

```diff
@@ -0,0 +1,194 @@
+import android.os.Handler;
+import android.provider.Settings;
+import android.view.KeyEvent;
+import android.widget.EditText;
+import android.widget.Toast;
+
+import com.facebook.common.logging.FLog;
+import com.facebook.react.common.ReactConstants;
+import com.facebook.react.modules.core.DefaultHardwareBackBtnHandler;
+import com.facebook.react.shell.MainReactPackage;
+
+import javax.annotation.Nullable;
+
+/**
+ * Base Activity for React Native applications.
+ */
+public abstract class ReactActivity extends Activity implements DefaultHardwareBackBtnHandler {
+
+    private static final String REDBOX_PERMISSION_MESSAGE =
+            "Overlay permissions needs to be granted in order for react native apps to run in dev mode";
+
+    private @Nullable ReactInstanceManager mReactInstanceManager;
+    private LifecycleState mLifecycleState = LifecycleState.BEFORE_RESUME;
+    private boolean mDoRefresh = false;
+
+    /**
+     * @return the name of the bundle in assets. If this is null, the app will only work with dev
+     * support on and will always try to load the JS bundle from the packager server.
+     * eg. "index.android.bundle"
+     */
+    protected abstract @Nullable String getBundleAssetName();
+
+    /**
+     * @return the path of the bundle file, if the bundle needs to be loaded from custom path
+     */
+    protected abstract @Nullable String getJSBundleFile();
+
+    /**
+     * @return the name of the main module. This is used to determine the URL to fetch the JS bundle
+     * from the packager server and is only used when dev support is enabled.
+     * This is the first file to be executed once the {@code ReactInstanceManager} is created.
+     * eg. "RKJSModules/Apps/Airbourne/AirbourneAppBundle"
+     */
+    protected abstract String getJSMainModuleName();
+
+    /**
+     * @return the name of the main component registered from JavaScript. This is used to schedule
+     * rendering of the component
+     */
+    protected abstract String getMainComponentName();
+
+    /**
+     * @return if debug mode should be enabled
+     */
+    protected abstract boolean getUseDeveloperSupport();
+
+
+    /**
+     * Method returns a {@link ReactPackage} associated with this class.
+     * A subclass may override this method if it needs more views or modules.
+     * In that case a subclass should return a {@link com.facebook.react.CompositeReactPackage} of
+     * this ReactPackage and its own.
+     */
+    protected @Nullable abstract ReactPackage getReactPackage();
+
+    /**
+     * @return a ReactInstanceManager
+     * A subclass may override this method if it needs to use a custom instance
+     */
+    protected ReactInstanceManager getReactInstanceManager() {
+        ReactInstanceManager.Builder builder = ReactInstanceManager.builder()
+                .setApplication(getApplication())
+                .setBundleAssetName(getBundleAssetName())
+                .setJSBundleFile(getJSBundleFile())
```

Comment:
I think you might have to call `setJSBundleFile` first? This way we'd revert it back to `null` after calling `setBundleAssetName`: https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/ReactInstanceManager.java#L183